src/liballoc_frame/lib.rs

```diff
@@ -0,0 +1,199 @@
+// constant at the call site and the branch will be optimized out.
+#[cfg(all(any(target_arch = "x86",
+              target_arch = "arm",
+              target_arch = "mips",
+              target_arch = "powerpc",
+              target_arch = "powerpc64",
+              target_arch = "asmjs",
+              target_arch = "wasm32")))]
+const MIN_ALIGN: usize = 8;
+#[cfg(all(any(target_arch = "x86_64",
+              target_arch = "aarch64",
+              target_arch = "mips64",
+              target_arch = "s390x")))]
+const MIN_ALIGN: usize = 16;
+
+// The size of the chunk which is actually allocated
+const CHUNK_SIZE: usize = 4096 * 16;
+const CHUNK_ALIGN: usize = 4096;
+
+static mut HEAP: *mut u8 = ptr::null_mut();
+static mut HEAP_LEFT: usize = 0;
+static HEAP_MUTEX: AtomicBool = AtomicBool::new(false);
+
+#[no_mangle]
+pub extern "C" fn __rust_allocate(size: usize, align: usize) -> *mut u8 {
+    let new_align = if align < MIN_ALIGN { MIN_ALIGN } else { align };
+    let new_size = (size + new_align - 1) & !(new_align - 1);
+
+    unsafe {
+        if new_size > CHUNK_SIZE {
+            return imp::allocate(size, align);
+        }
+        
+        while HEAP_MUTEX.compare_and_swap(false, true, Ordering::SeqCst) {}
+
+        if new_size < HEAP_LEFT {
+            let p = HEAP;
+            HEAP = p.offset(new_size as isize);
+            HEAP_LEFT -= new_size;
+            HEAP_MUTEX.store(false, Ordering::SeqCst);
+            return p;
+        } else {
+            let p = imp::allocate(CHUNK_SIZE, CHUNK_ALIGN);
+            HEAP = p.offset(new_size as isize);
+            HEAP_LEFT = CHUNK_SIZE - new_size;
+            HEAP_MUTEX.store(false, Ordering::SeqCst);
+            return p;
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn __rust_deallocate(_ptr: *mut u8, _old_size: usize, _align: usize) {
+}
+
+#[no_mangle]
+pub extern "C" fn __rust_reallocate(ptr: *mut u8,
+                                    old_size: usize,
+                                    size: usize,
+                                    align: usize)
+                                    -> *mut u8 {
+    let new_ptr = __rust_allocate(size, align);
+    unsafe { libc::memcpy(new_ptr as *mut _, ptr as *mut _, old_size); } 
+    new_ptr
+}
+
+#[no_mangle]
+pub extern "C" fn __rust_reallocate_inplace(_ptr: *mut u8,
+                                            old_size: usize,
+                                            _size: usize,
+                                            _align: usize)
+                                            -> usize {
+    old_size
+}
+
+#[no_mangle]
+pub extern "C" fn __rust_usable_size(size: usize, _align: usize) -> usize {
+    size
+}
+
+#[cfg(any(unix, target_os = "redox"))]
+mod imp {
+    use libc;
+    use core::ptr;
+    use MIN_ALIGN;
+
+    pub unsafe fn allocate(size: usize, align: usize) -> *mut u8 {
+        if align <= MIN_ALIGN {
+            libc::malloc(size as libc::size_t) as *mut u8
```

Comment:
Yes, this is probably a good idea. I'll look into using sbrk and VirtualPageAllocEx or whatever. The current implementation as simple as possible just to test out the idea.